brew_view/controllers/users_api.py

```diff
@@ -0,0 +1,244 @@
+        responses:
+          200:
+            description: User with the given ID
+            schema:
+              $ref: '#/definitions/User'
+          400:
+            $ref: '#/definitions/400Error'
+          404:
+            $ref: '#/definitions/404Error'
+          50x:
+            $ref: '#/definitions/50xError'
+        tags:
+          - Users
+        """
+        principal = Principal.objects.get(id=str(user_id))
+        operations = BeerGardenSchemaParser.parse_patch(
+            self.request.decoded_body,
+            many=True,
+            from_string=True
+        )
+
+        for op in operations:
+            if op.path == '/roles':
+                try:
+                    role = Role.objects.get(name=op.value)
+                except DoesNotExist:
+                    error_msg = "Role '%s' does not exist" % op.value
+                    self.logger.warning(error_msg)
+                    raise ModelValidationError(error_msg)
+
+                if op.operation == 'add':
+                    principal.roles.append(role)
+                elif op.operation == 'remove':
+                    principal.roles.remove(role)
+                else:
+                    error_msg = "Unsupported operation '%s'" % op.operation
+                    self.logger.warning(error_msg)
+                    raise ModelValidationError(error_msg)
+
+            elif op.path == '/permissions':
+                try:
+                    perm = Permissions(op.value)
+                except ValueError:
+                    error_msg = "Permission '%s' does not exist" % op.value
+                    self.logger.warning(error_msg)
+                    raise ModelValidationError(error_msg)
+
+                if op.operation == 'add':
+                    principal.permissions.append(perm.value)
+                elif op.operation == 'remove':
+                    principal.permissions.remove(perm.value)
+                else:
+                    error_msg = "Unsupported operation '%s'" % op.operation
+                    self.logger.warning(error_msg)
+                    raise ModelValidationError(error_msg)
+
+            elif op.path == '/preferences/theme':
+                if op.operation == 'set':
+                    principal.preferences['theme'] = op.value
+                else:
+                    error_msg = "Unsupported operation '%s'" % op.operation
+                    self.logger.warning(error_msg)
+                    raise ModelValidationError(error_msg)
+
+            else:
+                error_msg = "Unsupported path '%s'" % op.path
+                self.logger.warning(error_msg)
+                raise ModelValidationError(error_msg)
+
+        principal.save()
+
+        self.write(BeerGardenSchemaParser.serialize_principal(principal,
+                                                              to_string=False))
+
+
+class UsersAPI(BaseHandler):
+
+    @authenticated(permissions=[Permissions.USER_READ])
```

Comment:
It feels slightly weird that a person who can read one user can also read all users.

Reply:
Yeah, I need to clean this up a bit.